File: go/tests/travel_board_test.go
// travel_board_test.go
//
// Verifies the CircleAI.Travel port (travel_board.go): flight/stay/trip storage,
// trip cost totalling (flights + nightly*nights), and upcoming-trips ordering.

package circleai_test

import (
	"testing"
	"time"

	circleai "github.com/bhengubv/CircleAI/go"
)

func TestTravel_TripCost(t *testing.T) {
	b := circleai.NewInMemoryTravelBoard()
	b.AddFlight(circleai.Flight{FlightId: "f1", From: "JNB", To: "CPT", Price: circleai.DecimalFromInt(1200), Currency: "ZAR"})
	// 3 nights at 900 = 2700.
	ci := time.Date(2026, 8, 1, 0, 0, 0, 0, time.UTC)
	b.AddStay(circleai.HotelStay{StayId: "s1", Hotel: "Grand", City: "CPT", CheckIn: ci, CheckOut: ci.AddDate(0, 0, 3), NightlyRate: circleai.DecimalFromInt(900), Currency: "ZAR"})
	b.Plan(circleai.TravelTrip{TripId: "t1", Name: "Coast", StartDate: ci, EndDate: ci.AddDate(0, 0, 3), FlightIds: []string{"f1"}, StayIds: []string{"s1"}})

	if got, ok := b.GetTrip("t1"); !ok || got.Name != "Coast" {
		t.Fatalf("get trip = %+v ok=%v", got, ok)
	}
	if got, ok := b.GetFlight("f1"); !ok || got.To != "CPT" {
		t.Fatalf("get flight = %+v ok=%v", got, ok)
	}
	if got, ok := b.GetStay("s1"); !ok || got.Hotel != "Grand" {
		t.Fatalf("get stay = %+v ok=%v", got, ok)
	}

	cost, err := b.TripCost("t1")
	if err != nil {
		t.Fatalf("trip cost: %v", err)
	}
	if !cost.Equal(circleai.DecimalFromInt(3900)) {
		t.Fatalf("trip cost = %s, want 3900", cost.String())
	}
	if _, err := b.TripCost("ghost"); err == nil {
		t.Fatalf("unknown trip cost must error")
	}
}

func TestTravel_UpcomingTrips(t *testing.T) {
	b := circleai.NewInMemoryTravelBoard()
	now := time.Date(2026, 7, 1, 0, 0, 0, 0, time.UTC)
	b.Plan(circleai.TravelTrip{TripId: "t1", Name: "B", StartDate: now.AddDate(0, 0, 20)})
	b.Plan(circleai.TravelTrip{TripId: "t2", Name: "A", StartDate: now.AddDate(0, 0, 10)})
	b.Plan(circleai.TravelTrip{TripId: "t3", Name: "Past", StartDate: now.AddDate(0, 0, -10)})

	up := b.UpcomingTrips(now)
	if len(up) != 2 || up[0].TripId != "t2" || up[1].TripId != "t1" {
		t.Fatalf("upcoming trips ordered failed: %+v", up)
	}
}

File: typescript/tests/travel_board.test.ts
// travel_board.test.ts
// Verifies the CircleAI.Travel port: add flight/stay overloads, trip cost
// (flights + nights, min 1 night, missing ids skipped), upcoming trips.

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import {
  InMemoryTravelBoard,
  TravelDomainContext,
  flight,
  hotelStay,
  travelTrip,
} from "../src/travel/index";

describe("InMemoryTravelBoard", () => {
  it("computes trip cost from flight prices and hotel nights", () => {
    const b = new InMemoryTravelBoard();
    b.add(flight("f1", "JNB", "CPT", new Date("2026-02-01T08:00:00Z"), new Date("2026-02-01T10:00:00Z"), "CemAir", "Y", 1500, "ZAR"));
    b.add(hotelStay("s1", "Grand", "CPT", new Date("2026-02-01T00:00:00Z"), new Date("2026-02-04T00:00:00Z"), 1200, "ZAR"));
    b.plan(travelTrip("t1", "CT Trip", new Date("2026-02-01T00:00:00Z"), new Date("2026-02-04T00:00:00Z"), ["f1"], ["s1"]));
    // 1500 + 1200*3 nights = 5100
    assert.equal(b.tripCost("t1"), 5100);
    assert.equal(b.getFlight("f1")?.carrier, "CemAir");
    assert.equal(b.getStay("s1")?.hotel, "Grand");
  });

  it("charges a minimum of one night and skips missing flight/stay ids", () => {
    const b = new InMemoryTravelBoard();
    b.add(hotelStay("s1", "H", "X", new Date("2026-02-01T00:00:00Z"), new Date("2026-02-01T00:00:00Z"), 900, "ZAR")); // same-day → 1 night
    b.plan(travelTrip("t1", "T", new Date("2026-02-01T00:00:00Z"), new Date("2026-02-02T00:00:00Z"), ["missing"], ["s1"]));
    assert.equal(b.tripCost("t1"), 900);
  });

  it("tripCost throws on unknown trip", () => {
    const b = new InMemoryTravelBoard();
    assert.throws(() => b.tripCost("ghost"), /Unknown trip ghost/);
  });

  it("lists upcoming trips (StartDate >= now) ascending", () => {
    const b = new InMemoryTravelBoard();
    b.plan(travelTrip("t1", "Past", new Date("2026-01-01T00:00:00Z"), new Date("2026-01-02T00:00:00Z"), [], []));
    b.plan(travelTrip("t2", "Soon", new Date("2026-03-01T00:00:00Z"), new Date("2026-03-05T00:00:00Z"), [], []));
    b.plan(travelTrip("t3", "Later", new Date("2026-05-01T00:00:00Z"), new Date("2026-05-05T00:00:00Z"), [], []));
    assert.deepEqual(
      b.upcomingTrips(new Date("2026-02-01T00:00:00Z")).map((t) => t.tripId),
      ["t2", "t3"],
    );
  });

  it("domain context exposes prompt + compliance + tools", () => {
    assert.ok(TravelDomainContext.systemPromptSnippet.includes("[DOMAIN: Travel]"));
    assert.deepEqual(TravelDomainContext.complianceFlags, ["POPIA", "Consumer_Protection_Act", "IATA_aware"]);
    assert.deepEqual(TravelDomainContext.suggestedTools, ["flight_search", "mapping", "currency_converter", "web_search"]);
  });
});
